### PR TITLE
Migrate WinForms Matching game to VB

### PR DESCRIPTION
**Pull request by adegeo**
## Summary

- Migrated the Windows Forms Matching Game sample to VB.NET
- Added NuGet package `MetroFramework` to the original project to support the migration article's example copying NuGet packages.

### CODE REVIEW

**Review by BillWagner — APPROVED** [review]
This looks great @adegeo 

You can :shipit: when ready.

**Comment by merriemcgaw** [comment]
@KlausLoeffelmann you want to take a look?

File: windowsforms/matching-game/net45/vb/MatchingGame.Logic/Game.vb [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        Dim availableCards = New List(Of Char) From
+        {
+            "!"c, "!"c, "N"c, "N"c, ","c, ","c, "k"c, "k"c,
+            "b"c, "b"c, "v"c, "v"c, "w"c, "w"c, "z"c, "z"c
+        }
+
+        Dim random As New Random()
+
+        For i = 0 To numberOfPairs * 2 - 1
+            Dim cardIndex = random.Next(availableCards.Count - 1)
```
Comment (Youssef1313) [inline windowsforms/matching-game/net45/vb/MatchingGame.Logic/Game.vb:33]: The value passed to `Next` is exclusive. Was it meant that `Count - 1` is never generated?
(Applies to both C# and VB code)

File: windowsforms/matching-game/net5-windows/vb/MatchingGame.Logic/Game.vb [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        Next
+
+        Return New Game(width, height, board)
+    End Function
+
+    Public ReadOnly Property Turns As Integer
+        Get
+            Return _turns
+        End Get
+    End Property
```
Comment (Youssef1313) [inline windowsforms/matching-game/net5-windows/vb/MatchingGame.Logic/Game.vb:46]: Why this doesn't use auto-implemented property syntax like the `Width` and `Height` properties?

File: windowsforms/matching-game/net5-windows/vb/MatchingGame.Logic/Game.vb [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+    Private Function GetIndex(w As Integer, h As Integer) As Integer
+        Return h * Width + w
+    End Function
+
+    Public Function OpenCard(w As Integer, h As Integer) As Integer
+        If _turnCards.Count > 1 Then
+            Throw New InvalidOperationException("Cannot open more than two cards")
```
Comment (Youssef1313) [inline windowsforms/matching-game/net5-windows/vb/MatchingGame.Logic/Game.vb:62]: nit: Exception guidelines recommend a period at the end.

```suggestion
            Throw New InvalidOperationException("Cannot open more than two cards.")
```

File: windowsforms/matching-game/net5-windows/vb/MatchingGame.Logic/My Project/AssemblyInfo.vb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+﻿Imports System
```
Comment (Youssef1313) [inline windowsforms/matching-game/net5-windows/vb/MatchingGame.Logic/My Project/AssemblyInfo.vb:1]: Is AssemblyInfo needed?

File: windowsforms/matching-game/net5-windows/vb/MatchingGame/MainForm.vb [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                Dim label = DirectCast(tableLayoutPanel1.GetControlFromPosition(w, h), Label)
+                label.ForeColor = IIf(_game.IsOpen(w, h), Color.Black, label.BackColor)
+            Next
+        Next
+    End Sub
+
+    Private Sub label_Click(sender As Object, e As EventArgs) Handles label16.Click, label15.Click, label9.Click, label8.Click, label7.Click, label6.Click, label5.Click, label4.Click, label3.Click, label2.Click, label14.Click, label13.Click, label12.Click, label11.Click, label10.Click, label1.Click
+        If closeCardTimer.Enabled Then Return
+
+        If TypeOf sender Is Label Then
```
Comment (Youssef1313) [inline windowsforms/matching-game/net5-windows/vb/MatchingGame/MainForm.vb:35]: Can this ever be false?

File: windowsforms/matching-game/net5-windows/vb/MatchingGame/MainForm.vb [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    Private Sub UpdateCards()
+        For w = 0 To _game.Width - 1
+            For h = 0 To _game.Height - 1
+                Dim label = DirectCast(tableLayoutPanel1.GetControlFromPosition(w, h), Label)
+                label.ForeColor = IIf(_game.IsOpen(w, h), Color.Black, label.BackColor)
+            Next
+        Next
+    End Sub
+
+    Private Sub label_Click(sender As Object, e As EventArgs) Handles label16.Click, label15.Click, label9.Click, label8.Click, label7.Click, label6.Click, label5.Click, label4.Click, label3.Click, label2.Click, label14.Click, label13.Click, label12.Click, label11.Click, label10.Click, label1.Click
```
Comment (Youssef1313) [inline windowsforms/matching-game/net5-windows/vb/MatchingGame/MainForm.vb:32]: Re-ordering the labels helps to confirm that no label is missed when reading the code.

File: windowsforms/matching-game/net5-windows/vb/MigrationBackup/8e4640f7/MatchingGame/NuGetUpgradeLog.html [hunk 7]
```diff
@@ -0,0 +1,163 @@
+<!DOCTYPE html>
```
Comment (Youssef1313) [inline windowsforms/matching-game/net5-windows/vb/MigrationBackup/8e4640f7/MatchingGame/NuGetUpgradeLog.html:1]: Is this needed?

File: windowsforms/matching-game/net5-windows/vb/MigrationBackup/8e4640f7/MatchingGame/packages.config [hunk 8]
```diff
@@ -0,0 +1,7 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (Youssef1313) [inline windowsforms/matching-game/net5-windows/vb/MigrationBackup/8e4640f7/MatchingGame/packages.config:1]: Is this needed?

File: windowsforms/matching-game/net5-windows/vb/MigrationBackup/8e4640f7/MatchingGame/MatchingGame.vbproj [hunk 9]
```diff
@@ -0,0 +1,139 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (Youssef1313) [inline windowsforms/matching-game/net5-windows/vb/MigrationBackup/8e4640f7/MatchingGame/MatchingGame.vbproj:1]: Is this needed?

File: windowsforms/matching-game/net5-windows/vb/MatchingGame.Logic/My Project/AssemblyInfo.vb (hunk 4 above)
Comment (adegeo) [reply windowsforms/matching-game/net5-windows/vb/MatchingGame.Logic/My Project/AssemblyInfo.vb:1]: Yes, the migration article references this code and part of migration is keeping your assembly info.

File: windowsforms/matching-game/net5-windows/vb/MatchingGame.Logic/Game.vb (hunk 2 above)
Comment (adegeo) [reply windowsforms/matching-game/net5-windows/vb/MatchingGame.Logic/Game.vb:46]: I'm not sure. I didn't do a ton of cleanup on the original code. Considering this is supporting the migration article I don't think we'll modify this code unless there is a bug.

File: windowsforms/matching-game/net45/vb/MatchingGame.Logic/Game.vb (hunk 1 above)
Comment (adegeo) [reply windowsforms/matching-game/net45/vb/MatchingGame.Logic/Game.vb:33]: Unsure. I just used this code from a blog sample done by the team and then converted it. It plays correctly.

File: windowsforms/matching-game/net5-windows/vb/MatchingGame/MainForm.vb (hunk 5 above)
Comment (adegeo) [reply windowsforms/matching-game/net5-windows/vb/MatchingGame/MainForm.vb:35]: If code is added that calls it and passes null. Highly unlikely in this scenario. I think it's fine to keep.

**Comment by adegeo** [comment]
@Youssef1313 Thanks for the comments! I addressed a few in the code. Cheers!

**Pull request merged**